### PR TITLE
Rewrite Choice step to make it usable from mlrun

### PR DESCRIPTION
**Pull request by gtopper**
[ML-7818](https://iguazio.atlassian.net/browse/ML-7818)

[ML-7818]: https://iguazio.atlassian.net/browse/ML-7818?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: storey/flow.py [hunk 1]
```diff
@@ -347,46 +348,48 @@ def _get_uuid(self):
...
-
-        self._choice_array = choice_array
-        for outlet, _ in choice_array:
-            self.to(outlet)
-
-        if default:
-            self.to(default)
-        self._default = default
+    def _init(self, **kwargs):
+        super()._init()
```
Comment (guy1992l) [inline storey/flow.py:357]: Just to make sure it is not an error: before the change the `kwargs` were sent at the `Flow` init so I would expect to see `super().__init__(**kwargs)` unless it is redundant. I think the `outlets` param should be passed here?

File: storey/flow.py [hunk 2]
```diff
@@ -347,46 +348,48 @@ def _get_uuid(self):
...
+        self._name_to_outlet = {}
+        for outlet in self._outlets:
+            if outlet.name in self._name_to_outlet:
+                raise ValueError(f"Ambiguous outlet name '{outlet.name}' in Choice step")
+            self._name_to_outlet[outlet.name] = outlet
+        # TODO: hacky way of supporting mlrun preview, which replaces targets with a DFTarget
+        self._passthrough_for_preview = list(self._name_to_outlet) == ["dataframe"]
+
+    def select_outlets(self, event):
```
Comment (guy1992l) [inline storey/flow.py]: I would add a type hint and also in the doc string mention that the selected outlets should be returned in a list of strings - the outlet names.

```suggestion
    def select_outlets(self, event) -> List[str]:
```

File: storey/flow.py [hunk 3]
```diff
@@ -347,46 +348,48 @@ def _get_uuid(self):
...
-            await self._default._do(event)
+        if event is _termination_obj:
+            return await self._do_downstream(_termination_obj)
+        else:
+            event_body = event if self._full_event else event.body
+            outlet_names = self.select_outlets(event_body)
+            outlets = []
+            if self._passthrough_for_preview:
+                outlet = self._name_to_outlet["dataframe"]
+                outlets.append(outlet)
```
Comment (guy1992l) [inline storey/flow.py:382]: add a TODO here as well to remember to delete this "hack"

**Review by guy1992l — COMMENTED** [review]
I'm not the person who should review it, but if I may, it looks pretty much straight forward. I had a comment and few suggestions but besides these, A+ :)

File: storey/flow.py (hunk 3 above)
Comment (gtopper) [reply storey/flow.py:382]: I don't really want to duplicate the TODO that's already on the attribute definition.

File: storey/flow.py (hunk 1 above)
Comment (gtopper) [reply storey/flow.py:357]: Thanks, the `**kwargs` need to be removed. There's no kwargs in `_init()`. Note that `_init` is unrelated to `__init__`.

**Review by guy1992l — APPROVED** [review]

**Pull request merged**